### DISCUSSION TITLE
Creating custom mutation/query

### DISCUSSION THREAD

**Discussion by satvikt** [saleor/saleor#9203] [General]
What's the procedure/tasks required to create a **custom mutation** or **query** in Saleor? 
(I haven't been able to find a resource on the official docs or github discussions which provides this information)

From what I've explored, I'd have to 

- provide the schema under `saleor/schema.graphql`
- (for mutation) make a class subclassing `BaseMutation` under the appropriate package (like `saleor/graphql/account` or `saleor/graphql/checkout` etc.)

to name a few.

**Comment by krzysztofwolski** [comment]
Saleor uses Graphene to create gql API. This framework has code first approach (define classes and schema is generated based on it). Hope docs will help: https://docs.graphene-python.org/projects/django/en/latest/mutations/

**Comment by artemudovyk** [comment]
What's the recommended approach for this now, given we shouldn't fork?

**Comment by korycins** [comment]
Using appollo federation: https://docs.saleor.io/developer/extending/apollo-federation, https://www.apollographql.com/docs/federation/federated-schemas/

---

Or building the external service that would make your custom action, and then make an action (if needed) on Saleor side by using [app-token](https://docs.saleor.io/api-reference/apps/objects/app-create#appcreateauthtokenstring--)

---

Additionally, if you need to store one or two additional fields, you could also consider using `object.metadata` and/or `object.privateMetadata` - [docs](https://docs.saleor.io/api-usage/metadata)

**Comment by artemudovyk** [comment]
Thank you!
I found out about Apollo Federation support two days ago on Discord with the link to the documentation. It's exactly what I was looking for. :)
